internal/wal_segment_runner_test.go

```diff
@@ -27,40 +29,145 @@ func TestWalSegmentRunner_MissingSegments(t *testing.T) {
-	walSegments := make(map[internal.WalSegmentDescription]bool, 0)
-	for i := minSegmentNo; i <= maxSegmentNo; i++ {
+// TestWalSegmentRunner_SwitchTimelines tests that WalSegmentRunner correctly switch timelines
+func TestWalSegmentRunner_SwitchTimelines(t *testing.T) {
+	startTimelineId := uint32(2)
+	stopSegmentNo := internal.WalSegmentNo(1)
+	startSegment := internal.WalSegmentDescription{
+		Number:   internal.WalSegmentNo(8),
+		Timeline: startTimelineId,
+	}
+
+	storageSegmentsByTimeline := map[uint32][]internal.WalSegmentNo{
+		1: {
+			1, 2, 3, 4, 5, 6,
+		},
+		2: {
+			5, 6, 7,
+		},
+	}
+
+	expectedFoundSegmentsByTimeline := map[uint32][]internal.WalSegmentNo {
+		1: {
+			1, 2, 3, 4,
+		},
+		2: {
+			5, 6, 7,
+		},
+	}
+
+	timelineSwitchMap := map[internal.WalSegmentNo]*internal.TimelineHistoryRecord{
+		5: internal.NewTimelineHistoryRecord(1, 5 * internal.WalSegmentSize + 1, ""),
+	}
+
+	testWalSegmentRunnerMultipleTimelines(t, stopSegmentNo, startSegment, timelineSwitchMap,
+		storageSegmentsByTimeline, expectedFoundSegmentsByTimeline)
+}
+
+// TestWalSegmentRunner_SwitchTimelines tests that WalSegmentRunner correctly handles timeline switches
+// when missing segments exist
+func TestWalSegmentRunner_SwitchTimelinesMissingSegments(t *testing.T) {
+	startTimelineId := uint32(2)
+	stopSegmentNo := internal.WalSegmentNo(1)
+	startSegment := internal.WalSegmentDescription{
+		Number:   internal.WalSegmentNo(8),
+		Timeline: startTimelineId,
+	}
+
+	storageSegmentsByTimeline := map[uint32][]internal.WalSegmentNo{
+		1: {
+			1, 3, 4, 6,
+		},
+		2: {
+			6, 7,
+		},
+	}
+
+	expectedFoundSegmentsByTimeline := map[uint32][]internal.WalSegmentNo {
+		1: {
+			1, 3, 4,
+		},
+		2: {
+			6, 7,
+		},
+	}
+
+	timelineSwitchMap := map[internal.WalSegmentNo]*internal.TimelineHistoryRecord{
+		5: internal.NewTimelineHistoryRecord(1, 5 * internal.WalSegmentSize + 1, ""),
+	}
+
+	testWalSegmentRunnerMultipleTimelines(t, stopSegmentNo, startSegment, timelineSwitchMap,
+		storageSegmentsByTimeline, expectedFoundSegmentsByTimeline)
+}
+
+func testWalSegmentRunnerMultipleTimelines(t *testing.T, stopSegmentNo internal.WalSegmentNo,
+	startSegment internal.WalSegmentDescription, timelineSwitchMap map[internal.WalSegmentNo]*internal.TimelineHistoryRecord,
+	storageSegmentsByTimeline, expectedSegmentsByTimeline map[uint32][]internal.WalSegmentNo) {
+
+	// convert storageSegmentsByTimeline
+	storageSegments := make(map[internal.WalSegmentDescription]bool, 0)
+	for timeline, segmentNumbers := range storageSegmentsByTimeline {
+		for _, segmentNo := range segmentNumbers {
+			segment := internal.WalSegmentDescription{Number: segmentNo, Timeline: timeline}
+			storageSegments[segment] = true
+		}
+	}
+	walSegmentRunner := internal.NewWalSegmentRunner(startSegment, storageSegments, stopSegmentNo, timelineSwitchMap)
+
+	// convert expectedSegmentsByTimeline
+	expectedSegments := make(map[internal.WalSegmentDescription]bool, 0)
```

Comment:
// convert storageSegmentsByTimeline
and
// convert expectedSegmentsByTimeline
have the same code, DRY

Reply:
Fixed in https://github.com/wal-g/wal-g/pull/746/commits/b00b8d131ed0462e6aa017ea4771c094bedd41e2